firedrake/petsc.py

```diff
@@ -334,3 +370,70 @@ def garbage_view(obj: Any):
         PETSc.garbage_view(comm)
     else:
         raise FiredrakePETScError("No comm found, cannot view garbage")
+
+
+external_packages = get_external_packages()
+
+# Setup default partitioner
+# Manually define the priority until
+# https://petsc.org/main/src/dm/partitioner/interface/partitioner.c.html#PetscPartitionerGetDefaultType
+# is added to petsc4py
+partitioner_priority = ["parmetis", "ptscotch", "chaco"]
+for partitioner in partitioner_priority:
+    if partitioner in external_packages:
+        DEFAULT_PARTITIONER = partitioner
+        break
+else:
+    warn(
+        "No external package for " + ", ".join(partitioner_priority)
+        + " found, defaulting to PETSc simple partitioner. This may not be optimal."
+    )
+    DEFAULT_PARTITIONER = "simple"
+
+# Setup default direct solver
+direct_solver_priority = ["mumps", "superlu_dist", "pastix"]
+for solver in direct_solver_priority:
+    if solver in external_packages:
+        DEFAULT_DIRECT_SOLVER = solver
+        DEFAULT_DIRECT_SOLVER_PARAMETERS = {"mat_solver_type": solver}
+        break
+else:
+    warn(
+        "No external package for " + ", ".join(direct_solver_priority)
+        + " found, defaulting to PETSc LU. This will only work in serial."
+    )
+    DEFAULT_DIRECT_SOLVER = "petsc"
+    DEFAULT_DIRECT_SOLVER_PARAMETERS = {"mat_solver_type": "petsc"}
+
+# MUMPS needs an additional parameter set
+# From the MUMPS documentation:
+# > ICNTL(14) controls the percentage increase in the estimated working space...
+# > ... Remarks: When significant extra fill-in is caused by numerical pivoting, increasing ICNTL(14) may help.
+if DEFAULT_DIRECT_SOLVER == "mumps":
+    DEFAULT_DIRECT_SOLVER_PARAMETERS["mat_mumps_icntl_14"] = 200
+
+# Setup default AMG preconditioner
+amg_priority = ["hypre", "ml"]
+for amg in amg_priority:
+    if amg in external_packages:
+        DEFAULT_AMG_PC = amg
+else:
+    DEFAULT_AMG_PC = "gamg"
+
+
+DEFAULT_KSP_PARAMETERS = flatten_parameters({
```

Comment:
I think it's risky making these dictionaries module members like this. If they get modified accidentally (e.g. someone puts them into their own params dictionary then modifies that dictionary) then these defaults will then be different for any solvers created later.

Instead we could have a function to return a (new) dictionary with the default parameters. Essentially change from "here's our default dictionary" to "here's a dictionary containing our defaults".

Reply:
I have changed the code to use a `MappingProxyType` which essentially gives us an immutable `dict`. There are now private `_DEFAULT_KSP_PARAMETERS` and you could change these to globally change the defaults (if you wanted to do this for whatever reason) and public `DEFAULT_KSP_PARAMETERS = MappingProxyType(_DEFAULT_KSP_PARAMETERS)` which are immutable, negating the need for a copy whenever def DEFAULT is used.

Reply:
Actually, I make a deep copy of those private dicts. I don't really see a good case for globally editing the defaults from a script.

Reply:
Great, much nicer to just enforce immutability here.